File: src/lib/conversation-dag.ts
import {
  getConversation,
  getSubagentConversation,
  listLegacyConversations,
} from "./claude-data";
import type {
  ConversationDag,
  ConversationDagEdge,
  ConversationDagNode,
  ConversationDagStats,
  ConversationDagSummary,
  ProcessedConversation,
  SubagentInfo,
  SupportedProvider,
} from "./types";

function totalConversationTokens(conversation: ProcessedConversation): number {
  return (
    (conversation.totalUsage.input_tokens || 0) +
    (conversation.totalUsage.output_tokens || 0) +
    (conversation.totalUsage.cache_creation_input_tokens || 0) +
    (conversation.totalUsage.cache_read_input_tokens || 0)
  );
}

function summarizeConversationLabel(
  conversation: ProcessedConversation | null,
  fallback: string
): string {
  if (!conversation) {
    return fallback;
  }

  for (const message of conversation.messages) {
    for (const block of message.blocks) {
      if (block.type === "text" && block.text.trim()) {
        return block.text.trim().replace(/\s+/g, " ").slice(0, 120);
      }
    }
  }

  return fallback;
}

function buildConversationPath(projectPath: string, sessionId: string): string {
  return `/conversations/${encodeURIComponent(projectPath)}/${sessionId}`;
}

function toNode(
  projectPath: string,
  sessionId: string,
  depth: number,
  parentSessionId: string | undefined,
  conversation: ProcessedConversation | null,
  subagent: SubagentInfo | undefined,
  isRoot: boolean
): ConversationDagNode {
  const fallbackLabel =
    subagent?.nickname ||
    subagent?.description ||
    `${isRoot ? "Conversation" : "Sub-agent"} ${sessionId.slice(0, 8)}`;

  return {
    id: sessionId,
    sessionId,
    parentSessionId,
    projectPath,
    label: summarizeConversationLabel(conversation, fallbackLabel),
    description: subagent?.description,
    nickname: subagent?.nickname,
    subagentType: subagent?.subagentType,
    threadType: isRoot ? "main" : "subagent",
    hasTranscript: Boolean(conversation),
    model: conversation?.model,
    messageCount: conversation?.messages.length ?? 0,
    totalTokens: conversation ? totalConversationTokens(conversation) : 0,
    timestamp: conversation?.lastUpdatedAt ?? 0,
    depth,
    path: buildConversationPath(projectPath, sessionId),
    isRoot,
  };
}

function computeStats(
  rootSessionId: string,
  nodes: ConversationDagNode[],
  edges: ConversationDagEdge[]
): ConversationDagStats {
  const breadthMap = new Map<number, number>();
  const outDegree = new Map<string, number>();

  for (const node of nodes) {
    breadthMap.set(node.depth, (breadthMap.get(node.depth) || 0) + 1);
    outDegree.set(node.id, 0);
  }

  for (const edge of edges) {
    outDegree.set(edge.source, (outDegree.get(edge.source) || 0) + 1);
  }

  const maxBreadth = Math.max(0, ...breadthMap.values());
  const maxDepth = Math.max(0, ...nodes.map((node) => node.depth));
  const leafCount = nodes.filter((node) => (outDegree.get(node.id) || 0) === 0).length;
  const rootSubagentCount = outDegree.get(rootSessionId) || 0;

  return {
    totalNodes: nodes.length,
    totalEdges: edges.length,
    totalSubagentNodes: nodes.filter((node) => !node.isRoot).length,
    maxDepth,
    maxBreadth,
    leafCount,
    rootSubagentCount,
    totalMessages: nodes.reduce((sum, node) => sum + node.messageCount, 0),
    totalTokens: nodes.reduce((sum, node) => sum + node.totalTokens, 0),
  };
}

export async function buildConversationDag(
  projectPath: string,
  rootSessionId: string
): Promise<ConversationDag | null> {
  const nodes = new Map<string, ConversationDagNode>();
  const edges = new Map<string, ConversationDagEdge>();
  const expanded = new Set<string>();
  const activePath = new Set<string>();

  async function visitThread(
    sessionId: string,
    depth: number,
    parentSessionId?: string,
    subagent?: SubagentInfo
  ): Promise<void> {
    if (parentSessionId) {
      const edgeId = `${parentSessionId}->${sessionId}`;
      edges.set(edgeId, {
        id: edgeId,
        source: parentSessionId,
        target: sessionId,
      });
    }

    const existingNode = nodes.get(sessionId);
    if (existingNode && existingNode.depth <= depth && expanded.has(sessionId)) {
      return;
    }

    const shouldLoadConversation = subagent?.hasFile !== false || !subagent;
    const conversation = shouldLoadConversation
      ? parentSessionId
        ? await getSubagentConversation(projectPath, parentSessionId, sessionId)
        : await getConversation(projectPath, sessionId)
      : null;
    const nextNode = toNode(
      projectPath,
      sessionId,
      existingNode ? Math.min(existingNode.depth, depth) : depth,
      parentSessionId,
      conversation,
      subagent,
      !parentSessionId
    );
    nodes.set(sessionId, existingNode ? { ...existingNode, ...nextNode } : nextNode);

    if (!conversation || expanded.has(sessionId) || activePath.has(sessionId)) {
      return;
    }

    activePath.add(sessionId);
    expanded.add(sessionId);

    for (const child of conversation.subagents) {
      if (child.agentId === sessionId) {
        continue;
      }
      await visitThread(child.agentId, nextNode.depth + 1, sessionId, child);
    }

    activePath.delete(sessionId);
  }

  await visitThread(rootSessionId, 0);

  const rootNode = nodes.get(rootSessionId);
  if (!rootNode) {
    return null;
  }

  const orderedNodes = Array.from(nodes.values()).sort((a, b) => {
    if (a.depth !== b.depth) {
      return a.depth - b.depth;
    }
    return (a.timestamp || 0) - (b.timestamp || 0) || a.id.localeCompare(b.id);
  });
  const orderedEdges = Array.from(edges.values()).sort((a, b) =>
    a.id.localeCompare(b.id)
  );

  return {
    projectPath,
    rootSessionId,
    nodes: orderedNodes,
    edges: orderedEdges,
    stats: computeStats(rootSessionId, orderedNodes, orderedEdges),
  };
}

export async function listConversationDagSummaries(
  projectFilter?: string,
  days?: number,
  provider?: SupportedProvider
): Promise<ConversationDagSummary[]> {
  const conversations = await listLegacyConversations(projectFilter, days);
  const mainsWithSubagents = conversations.filter((conversation) => {
    if (conversation.threadType !== "main" || conversation.subagentCount === 0) {
      return false;
    }
    if (provider === "codex") {
      return conversation.projectPath.startsWith("codex:");
    }
    if (provider === "claude") {
      return !conversation.projectPath.startsWith("codex:");
    }
    return true;
  });

  const summaries = await Promise.all(
    mainsWithSubagents.map(async (conversation) => {
      const dag = await buildConversationDag(
        conversation.projectPath,
        conversation.sessionId
      );
      if (!dag) {
        return null;
      }
      return {
        ...conversation,
        dag: dag.stats,
      } satisfies ConversationDagSummary;
    })
  );

  return summaries
    .filter((summary): summary is ConversationDagSummary => summary !== null)
    .sort((a, b) => b.lastUpdatedAt - a.lastUpdatedAt);
}

File: python/helaicopter_api/pure/conversation_dag.py
"""Pure conversation DAG construction for backend APIs."""

from __future__ import annotations

from collections.abc import Callable
from urllib.parse import quote

from helaicopter_domain.ids import SessionId
from helaicopter_api.schema.conversations import (
    ConversationDagEdgeResponse,
    ConversationDagNodeResponse,
    ConversationDagResponse,
    ConversationDagStatsResponse,
    ConversationDetailResponse,
    ConversationSubagentResponse,
    ConversationTextBlockResponse,
)


def build_conversation_dag(
    *,
    project_path: str,
    root_session_id: SessionId,
    load_conversation: Callable[[SessionId, SessionId | None], ConversationDetailResponse | None],
) -> ConversationDagResponse | None:
    """Build a deterministic conversation/sub-agent DAG from shaped details."""
    nodes: dict[str, ConversationDagNodeResponse] = {}
    edges: dict[str, ConversationDagEdgeResponse] = {}
    expanded: set[str] = set()
    active_path: set[str] = set()

    def visit(
        session_id: SessionId,
        depth: int,
        *,
        parent_session_id: SessionId | None = None,
        subagent: ConversationSubagentResponse | None = None,
    ) -> None:
        if parent_session_id is not None:
            edge_id = f"{parent_session_id}->{session_id}"
            edges[edge_id] = ConversationDagEdgeResponse(
                id=edge_id,
                source=parent_session_id,
                target=session_id,
            )

        existing = nodes.get(session_id)
        if existing is not None and existing.depth <= depth and session_id in expanded:
            return

        should_load = subagent is None or subagent.has_file
        conversation = load_conversation(session_id, parent_session_id) if should_load else None
        next_node = _to_node(
            project_path=project_path,
            session_id=session_id,
            depth=min(existing.depth, depth) if existing is not None else depth,
            parent_session_id=parent_session_id,
            conversation=conversation,
            subagent=subagent,
            is_root=parent_session_id is None,
        )
        nodes[session_id] = next_node if existing is None else existing.model_copy(update=next_node.model_dump())

        if conversation is None or session_id in expanded or session_id in active_path:
            return

        active_path.add(session_id)
        expanded.add(session_id)
        for child in conversation.subagents:
            if child.agent_id == session_id:
                continue
            visit(
                _child_session_id(child),
                next_node.depth + 1,
                parent_session_id=session_id,
                subagent=child,
            )
        active_path.remove(session_id)

    visit(root_session_id, 0)
    root_node = nodes.get(root_session_id)
    if root_node is None or not root_node.has_transcript or root_node.message_count == 0:
        return None

    ordered_nodes = sorted(
        nodes.values(),
        key=lambda node: (node.depth, node.timestamp, node.id),
    )
    ordered_edges = sorted(edges.values(), key=lambda edge: edge.id)
    return ConversationDagResponse(
        project_path=project_path,
        root_session_id=root_session_id,
        nodes=ordered_nodes,
        edges=ordered_edges,
        stats=_compute_stats(root_session_id=root_session_id, nodes=ordered_nodes, edges=ordered_edges),
    )


def _to_node(
    *,
    project_path: str,
    session_id: SessionId,
    depth: int,
    parent_session_id: SessionId | None,
    conversation: ConversationDetailResponse | None,
    subagent: ConversationSubagentResponse | None,
    is_root: bool,
) -> ConversationDagNodeResponse:
    fallback_label = (
        (subagent.nickname if subagent is not None else None)
        or (subagent.description if subagent is not None else None)
        or f"{'Conversation' if is_root else 'Sub-agent'} {session_id[:8]}"
    )
    return ConversationDagNodeResponse(
        id=session_id,
        session_id=session_id,
        parent_session_id=parent_session_id,
        project_path=project_path,
        label=_summarize_label(conversation, fallback_label),
        description=subagent.description if subagent is not None else None,
        nickname=subagent.nickname if subagent is not None else None,
        subagent_type=subagent.subagent_type if subagent is not None else None,
        thread_type="main" if is_root else "subagent",
        has_transcript=conversation is not None,
        model=conversation.model if conversation is not None else None,
        message_count=len(conversation.messages) if conversation is not None else 0,
        total_tokens=_conversation_total_tokens(conversation),
        timestamp=conversation.last_updated_at if conversation is not None else 0,
        depth=depth,
        path=_conversation_node_path(
            conversation_ref=(
                conversation.conversation_ref
                if conversation is not None
                else subagent.conversation_ref if subagent is not None else None
            ),
        ),
        is_root=is_root,
    )


def _conversation_node_path(
    *,
    conversation_ref: str | None,
) -> str | None:
    if not conversation_ref:
        return None
    return f"/conversations/by-ref/{quote(conversation_ref, safe='')}"


def _summarize_label(conversation: ConversationDetailResponse | None, fallback: str) -> str:
    if conversation is None:
        return fallback
    for message in conversation.messages:
        for block in message.blocks:
            if isinstance(block, ConversationTextBlockResponse) and block.text and block.text.strip():
                return " ".join(block.text.split())[:120]
    return fallback


def _conversation_total_tokens(conversation: ConversationDetailResponse | None) -> int:
    if conversation is None:
        return 0
    usage = conversation.total_usage
    if conversation.provider == "codex":
        return usage.input_tokens + usage.output_tokens
    return (
        usage.input_tokens
        + usage.output_tokens
        + usage.cache_creation_tokens
        + usage.cache_read_tokens
    )


def _child_session_id(child: ConversationSubagentResponse) -> SessionId:
    return SessionId(child.agent_id)


def _compute_stats(
    *,
    root_session_id: SessionId,
    nodes: list[ConversationDagNodeResponse],
    edges: list[ConversationDagEdgeResponse],
) -> ConversationDagStatsResponse:
    breadth_by_depth: dict[int, int] = {}
    out_degree: dict[str, int] = {}

    for node in nodes:
        breadth_by_depth[node.depth] = breadth_by_depth.get(node.depth, 0) + 1
        out_degree[node.id] = 0
    for edge in edges:
        out_degree[edge.source] = out_degree.get(edge.source, 0) + 1

    return ConversationDagStatsResponse(
        total_nodes=len(nodes),
        total_edges=len(edges),
        total_subagent_nodes=sum(1 for node in nodes if not node.is_root),
        max_depth=max((node.depth for node in nodes), default=0),
        max_breadth=max(breadth_by_depth.values(), default=0),
        leaf_count=sum(1 for node in nodes if out_degree.get(node.id, 0) == 0),
        root_subagent_count=out_degree.get(root_session_id, 0),
        total_messages=sum(node.message_count for node in nodes),
        total_tokens=sum(node.total_tokens for node in nodes),
    )
